### DISCUSSION TITLE
cpp/basics/const

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#27] [Announcements]
# cpp/basics/const

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/basics/const.html

**Comment by treble-h** [comment]
const int b = func(); // b 的值为 10，不能被修改
这里不能被修改是因为使用了const 修饰了b;
int b = func(); 实际是可以修改的, 等同于const a; int b = a; 将一个常量的值赋给了一个非常量

**Reply by whw-create** [reply]
所以我感觉const修饰函数返回值好像也没啥用

**Reply by huangjiawei128** [reply]
但是const修饰的函数返回值如果是一个指针就有意义了，例如：
const int* func1() {
    int* a = new int;
    *a = 128;
    return a;
}

int main() {
    int *b = func1();
}
会编译失败，因为const int*通常无法隐式的转换为int*，除非变成int *b = const_cast<int *>func1();或者const int *b = func1();

**Comment by zhi-lu** [comment]
想不让编译器做优化，让 10 变成 20给 a 再定义一个 volatile 关键字。

**Comment by huangjiawei128** [comment]
局部 const 变量在栈上分配地址。
全局或静态 const 变量在数据段（通常是.rodata段）分配地址。
constexpr 变量可能不会分配地址，取决于编译器优化。
类的 const 成员变量会在类实例内分配地址。
使用 extern 的 const 变量在定义文件中分配地址，在声明文件中不分配地址。
